File: Front-End/lib/barcode/nomenclature/barcode_rule.dart
/// The company's barcode nomenclature.
///
/// Mirrors `Back-End/Web-POS.Api/Domain/BarcodeRule.cs` and
/// `Services/BarcodeRuleMatcher.cs`. The POS must decode a scale label while
/// offline, so the matching logic lives on both sides rather than behind an
/// endpoint — keep the two implementations in step.
library;

/// What the value embedded in a matched barcode MEANS.
enum BarcodeRuleType {
  /// Plain product barcode — no embedded value.
  unit,

  /// Embedded value is a quantity in the product's own unit.
  weighted,

  /// Embedded value is a line TOTAL; quantity = value / unit price.
  priced,

  /// Embedded value is a percentage discount for the line.
  discounted,
}

/// Symbology the rule is restricted to. [BarcodeEncoding.any] skips length and
/// check-digit validation entirely.
enum BarcodeEncoding { any, ean13, upcA }

BarcodeRuleType _typeFromApi(String? raw) => switch (raw?.toLowerCase()) {
      'weighted' => BarcodeRuleType.weighted,
      'priced' => BarcodeRuleType.priced,
      'discounted' => BarcodeRuleType.discounted,
      _ => BarcodeRuleType.unit,
    };

String typeToApi(BarcodeRuleType type) => switch (type) {
      BarcodeRuleType.weighted => 'Weighted',
      BarcodeRuleType.priced => 'Priced',
      BarcodeRuleType.discounted => 'Discounted',
      BarcodeRuleType.unit => 'Unit',
    };

BarcodeEncoding _encodingFromApi(String? raw) => switch (raw?.toLowerCase()) {
      'ean13' => BarcodeEncoding.ean13,
      'upca' => BarcodeEncoding.upcA,
      _ => BarcodeEncoding.any,
    };

String encodingToApi(BarcodeEncoding encoding) => switch (encoding) {
      BarcodeEncoding.ean13 => 'Ean13',
      BarcodeEncoding.upcA => 'UpcA',
      BarcodeEncoding.any => 'Any',
    };

/// Human label for the encoding column.
String encodingLabel(BarcodeEncoding encoding) => switch (encoding) {
      BarcodeEncoding.ean13 => 'EAN-13',
      BarcodeEncoding.upcA => 'UPC-A',
      BarcodeEncoding.any => 'Any',
    };

/// One line of the nomenclature.
class BarcodeRule {
  final int id;
  final String name;

  /// Ascending evaluation order — first match wins.
  final int sequence;

  final BarcodeRuleType type;
  final BarcodeEncoding encoding;

  /// Prefix pattern, e.g. `22.....{NNDDD}`.
  final String pattern;

  final bool isEnabled;

  const BarcodeRule({
    required this.id,
    required this.name,
    required this.sequence,
    required this.type,
    required this.encoding,
    required this.pattern,
    this.isEnabled = true,
  });

  factory BarcodeRule.fromJson(Map<String, dynamic> json) => BarcodeRule(
        id: (json['id'] as num?)?.toInt() ?? 0,
        name: json['name'] as String? ?? '',
        sequence: (json['sequence'] as num?)?.toInt() ?? 0,
        type: _typeFromApi(json['type'] as String?),
        encoding: _encodingFromApi(json['encoding'] as String?),
        pattern: json['pattern'] as String? ?? '',
        isEnabled: json['isEnabled'] as bool? ?? true,
      );

  Map<String, dynamic> toJson() => {
        'id': id,
        'name': name,
        'sequence': sequence,
        'type': typeToApi(type),
        'encoding': encodingToApi(encoding),
        'pattern': pattern,
        'isEnabled': isEnabled,
      };

  BarcodeRule copyWith({
    String? name,
    int? sequence,
    BarcodeRuleType? type,
    BarcodeEncoding? encoding,
    String? pattern,
    bool? isEnabled,
  }) =>
      BarcodeRule(
        id: id,
        name: name ?? this.name,
        sequence: sequence ?? this.sequence,
        type: type ?? this.type,
        encoding: encoding ?? this.encoding,
        pattern: pattern ?? this.pattern,
        isEnabled: isEnabled ?? this.isEnabled,
      );
}

/// Outcome of matching a scanned barcode against a nomenclature.
class BarcodeMatch {
  final BarcodeRule rule;

  /// The barcode with the embedded-value digits blanked back to zeros — this is
  /// what the product's stored barcode must equal.
  final String productKey;

  /// The decoded embedded number, already scaled by its decimal positions.
  /// Zero for [BarcodeRuleType.unit] rules.
  final double value;

  const BarcodeMatch({
    required this.rule,
    required this.productKey,
    required this.value,
  });
}

/// The default nomenclature, used when the company has none yet (a fresh install
/// that has not synced, or an offline first run). Mirrors `BarcodeRuleSeeder`.
///
/// Order matters: the catch-all Unit rule must stay last, or it swallows every
/// scale label before the weighted rule is ever consulted.
const List<BarcodeRule> kDefaultBarcodeRules = [
  BarcodeRule(
      id: -1,
      name: 'Price Barcodes 2 Decimals',
      sequence: 10,
      type: BarcodeRuleType.priced,
      encoding: BarcodeEncoding.ean13,
      pattern: '25.....{NNNDD}'),
  BarcodeRule(
      id: -2,
      name: 'Weight Barcodes 3 Decimals',
      sequence: 20,
      type: BarcodeRuleType.weighted,
      encoding: BarcodeEncoding.ean13,
      pattern: '22.....{NNDDD}'),
  BarcodeRule(
      id: -3,
      name: 'Discount Barcodes',
      sequence: 30,
      type: BarcodeRuleType.discounted,
      encoding: BarcodeEncoding.any,
      pattern: '22{NN}'),
  BarcodeRule(
      id: -4,
      name: 'Product Barcodes',
      sequence: 40,
      type: BarcodeRuleType.unit,
      encoding: BarcodeEncoding.any,
      pattern: '.*'),
];

File: Back-End/Web-POS.Api/Domain/BarcodeRule.cs
using System;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace Api.Domain
{
    /// <summary>
    /// What the value embedded in a matched barcode MEANS.
    /// </summary>
    public enum BarcodeRuleType
    {
        /// <summary>Plain product barcode — no embedded value.</summary>
        Unit = 0,

        /// <summary>Embedded value is a quantity in the product's own unit.</summary>
        Weighted = 1,

        /// <summary>Embedded value is a line TOTAL; quantity = value / unit price.</summary>
        Priced = 2,

        /// <summary>Embedded value is a percentage discount for the line.</summary>
        Discounted = 3
    }

    /// <summary>
    /// Symbology the rule is restricted to. <see cref="Any"/> skips length and
    /// check-digit validation entirely.
    /// </summary>
    public enum BarcodeEncoding
    {
        Any = 0,
        Ean13 = 1,
        UpcA = 2
    }

    /// <summary>
    /// One line of the company's barcode nomenclature.
    /// </summary>
    /// <remarks>
    /// Rules are evaluated in <see cref="Sequence"/> order and the FIRST whose
    /// <see cref="Pattern"/> matches wins — which is why the catch-all
    /// <c>.*</c> Unit rule must always sort last. See
    /// <c>Services/BarcodeRuleMatcher.cs</c> for the pattern syntax.
    ///
    /// This replaces the six single-rule <c>Scale.Barcode.*</c> application
    /// properties, which could only ever express one scale format per company.
    /// </remarks>
    [Table("BarcodeRule")]
    public class BarcodeRule
    {
        [Key]
        public int Id { get; private set; }

        public int CompanyId { get; private set; }

        [Required, MaxLength(100)]
        public string Name { get; private set; } = default!;

        /// <summary>Ascending evaluation order. Ties broken by <see cref="Id"/>.</summary>
        public int Sequence { get; private set; }

        public BarcodeRuleType Type { get; private set; }

        public BarcodeEncoding Encoding { get; private set; }

        /// <summary>
        /// Prefix pattern, e.g. <c>22.....{NNDDD}</c>. Digits match themselves,
        /// <c>.</c> matches any character, and the <c>{N…D…}</c> group marks the
        /// embedded numeric value with <c>D</c> positions as decimals.
        /// </summary>
        [Required, MaxLength(100)]
        public string Pattern { get; private set; } = default!;

        public bool IsEnabled { get; private set; } = true;

        public DateTime LastModified { get; private set; } = DateTime.UtcNow;

        [ForeignKey(nameof(CompanyId))]
        public virtual Company? Company { get; private set; }

        public BarcodeRule() { }

        private BarcodeRule(int companyId, string name, int sequence, BarcodeRuleType type,
                            BarcodeEncoding encoding, string pattern, bool isEnabled)
        {
            CompanyId = companyId;
            Name = name;
            Sequence = sequence;
            Type = type;
            Encoding = encoding;
            Pattern = pattern;
            IsEnabled = isEnabled;
            LastModified = DateTime.UtcNow;
        }

        public static BarcodeRule Create(int companyId, string name, int sequence, BarcodeRuleType type,
                                         BarcodeEncoding encoding, string pattern, bool isEnabled = true)
        {
            if (companyId <= 0) throw new ArgumentException("CompanyId must be valid.", nameof(companyId));
            if (string.IsNullOrWhiteSpace(name)) throw new ArgumentException("Rule name cannot be empty.", nameof(name));
            if (string.IsNullOrWhiteSpace(pattern)) throw new ArgumentException("Pattern cannot be empty.", nameof(pattern));

            return new BarcodeRule(companyId, name.Trim(), sequence, type, encoding, pattern.Trim(), isEnabled);
        }

        public void Update(string? name, int? sequence, BarcodeRuleType? type,
                           BarcodeEncoding? encoding, string? pattern, bool? isEnabled)
        {
            if (!string.IsNullOrWhiteSpace(name)) Name = name.Trim();
            if (sequence.HasValue) Sequence = sequence.Value;
            if (type.HasValue) Type = type.Value;
            if (encoding.HasValue) Encoding = encoding.Value;
            if (!string.IsNullOrWhiteSpace(pattern)) Pattern = pattern.Trim();
            if (isEnabled.HasValue) IsEnabled = isEnabled.Value;

            LastModified = DateTime.UtcNow;
        }
    }
}
